quickwit/quickwit-indexing/src/source/queue_sources/processor.rs

```diff
@@ -0,0 +1,464 @@
+            self.observable_state.num_consecutive_empty_batches += 1;
+            return Ok(());
+        } else {
+            self.observable_state.num_consecutive_empty_batches = 0;
+        }
+
+        let preprocessed_messages = raw_messages
+            .into_iter()
+            .map(|m| m.pre_process(self.queue_params.message_type))
+            .collect::<anyhow::Result<Vec<_>>>()?;
+
+        let categorized_using_local_state =
+            self.local_state.filter_completed(preprocessed_messages);
+
+        let categorized_using_shared_state = self
+            .shared_state
+            .checkpoint_messages(
+                &self.publish_token,
+                categorized_using_local_state.processable,
+            )
+            .await?;
+
+        // Drop visibility tasks for messages that have been processed by another pipeline
+        let completed_visibility_tasks = categorized_using_shared_state
+            .already_processed
+            .iter()
+            .filter_map(|m| self.local_state.mark_completed(m.partition_id()))
+            .collect();
+        acknowledge(&*self.queue, completed_visibility_tasks).await?;
+
+        // Acknowledge messages that have been processed by another pipeline
+        let mut already_processed = categorized_using_local_state.already_processed;
+        already_processed.extend(categorized_using_shared_state.already_processed);
+        acknowledge(&*self.queue, already_processed).await?;
+
+        let ready_messages =
+            self.spawn_visibility_tasks(categorized_using_shared_state.processable);
+
+        self.local_state.set_ready_messages(ready_messages);
+
+        Ok(())
+    }
+
+    pub async fn emit_batches(
+        &mut self,
+        doc_processor_mailbox: &Mailbox<DocProcessor>,
+        ctx: &SourceContext,
+    ) -> Result<Duration, ActorExitStatus> {
+        if let Some(msg) = self.local_state.in_progress_mut() {
+            debug!("process_batch");
+            msg.process_batch(doc_processor_mailbox, ctx).await?;
+            if msg.is_eof() {
+                self.local_state.set_in_progress(None);
+            }
+        } else if let Some(ready_message) = self.local_state.get_ready_message() {
+            debug!("start_processing");
+            let new_in_progress = ready_message
+                .start_processing(&self.storage_resolver, self.source_type)
+                .await?;
+            self.local_state.set_in_progress(new_in_progress);
+        } else {
+            debug!("poll_messages");
+            self.poll_messages(ctx).await?;
+        }
+
+        Ok(Duration::ZERO)
+    }
+
+    pub async fn suggest_truncate(
+        &mut self,
+        checkpoint: SourceCheckpoint,
+        _ctx: &SourceContext,
+    ) -> anyhow::Result<()> {
+        let completed_visibility_handles = checkpoint
+            .iter()
+            .filter(|(_, pos)| pos.is_eof())
+            .filter_map(|(pid, _)| self.local_state.mark_completed(pid))
```

Comment:
Using the fluent iterator style to mutate stuff make things confusing
(here `mark_completed`)  and is error prone.

side effects like logging etc are ok of course. But modifying an object etc. is to avoid.
Prefer a simple look here.